pkg/dns/treecache.go

```diff
@@ -18,108 +18,68 @@ package dns
 
 import (
 	"bytes"
-	"crypto/md5"
 	"encoding/json"
-	"fmt"
-	"io/ioutil"
-	"os"
-	"path"
-	"reflect"
 	"strings"
 	"sync"
 )
 
-const (
-	dataFile = "data.dat"
-	crcFile  = "data.crc"
-)
-
 type object interface{}
 
 type TreeCache struct {
 	ChildNodes map[string]*TreeCache
 	Entries    map[string]interface{}
-	m          *sync.RWMutex
+	rwLock     *sync.RWMutex
 }
 
-func NewTreeCache() *TreeCache {
+func NewTreeCache(m *sync.RWMutex) *TreeCache {
 	return &TreeCache{
 		ChildNodes: make(map[string]*TreeCache),
 		Entries:    make(map[string]interface{}),
-		m:          &sync.RWMutex{},
+		rwLock:     m,
 	}
 }
 
-func Deserialize(dir string) (*TreeCache, error) {
-	b, err := ioutil.ReadFile(path.Join(dir, dataFile))
-	if err != nil {
-		return nil, err
-	}
-	var hash []byte
-	hash, err = ioutil.ReadFile(path.Join(dir, crcFile))
-	if err != nil {
-		return nil, err
-	}
-	if !reflect.DeepEqual(hash, getMD5(b)) {
-		return nil, fmt.Errorf("Checksum failed")
+func (cache *TreeCache) Serialize() (string, error) {
+	if cache.rwLock != nil {
+		cache.rwLock.RLock()
+		defer cache.rwLock.RUnlock()
 	}
-
-	var cache TreeCache
-	err = json.Unmarshal(b, &cache)
-	if err != nil {
-		return nil, err
-	}
-	cache.m = &sync.RWMutex{}
-	return &cache, nil
-}
-
-func (cache *TreeCache) Serialize(dir string) (string, error) {
-	cache.m.RLock()
-	defer cache.m.RUnlock()
 	b, err := json.Marshal(cache)
 	if err != nil {
 		return "", err
 	}
 
-	if len(dir) == 0 {
-		var prettyJSON bytes.Buffer
-		err = json.Indent(&prettyJSON, b, "", "\t")
+	var prettyJSON bytes.Buffer
+	err = json.Indent(&prettyJSON, b, "", "\t")
 
-		if err != nil {
-			return "", err
-		}
-		return string(prettyJSON.Bytes()), nil
-	}
-	if err := ensureDir(dir, os.FileMode(0755)); err != nil {
-		return "", err
-	}
-	if err := ioutil.WriteFile(path.Join(dir, dataFile), b, 0644); err != nil {
-		return "", err
-	}
-	if err := ioutil.WriteFile(path.Join(dir, crcFile), getMD5(b), 0644); err != nil {
+	if err != nil {
 		return "", err
 	}
-	return string(b), nil
+	return string(prettyJSON.Bytes()), nil
 }
 
-func (cache *TreeCache) SetEntry(key string, val interface{}, path ...string) {
-	cache.m.Lock()
-	defer cache.m.Unlock()
+func (cache *TreeCache) setEntry(key string, val interface{}, path ...string) {
+	if cache.rwLock != nil {
```

Comment:
as far as I can tell, this is only ever called on disconnected trees.  just skip the lock here and comment that this function assumes the tree is locked.